### PR TITLE
feat: support ref for Table

### LINKED ISSUE #25306: antd 4.x Table 不支持ref
- [ ] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[https://ant.design/components/table-cn/#API](https://ant.design/components/table-cn/#API)

### Steps to reproduce
/** @desc 类组件直接引用即可 */
import React from 'react';
import {Table} from 'antd';

type IProps = ReturnType<typeof mapStateToProps>;
const mapStateToProps = (rootState: IRootState) => ({
  global: rootState.Common_global,
});

Class TableAutoFixed extends Component<IProps,{}> {
   const tableRef:any = React.createRef();
   ...
   return <Table ref={this.tableRef}/>;
}
export default connect(mapStateToProps, null, null, { forwardRef: true })(TableAutoFixed);

### What is expected?
ref正常使用

### What is actually happening?
通过ref写的功能无法正常使用，ts检查的时候报错Table无ref属性，同时使用了react-redux的connect功能，虽然已经采用透传forwardRef方式:connect(mapStateToProps, null, null, { forwardRef: true })(TableAutoFixed)，但仍报错 Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()

| Environment | Info |
|---|---|
| antd | 4.0.0 |
| React | 16.8.6 |
| System | mac  |
| Browser | chrome |

---
TableAutoFixed 是实现表格依据可视窗口大小来判断是否对table固定列的功能

### PR DESCRIPTION
**Pull request by JounQin**
close #25306

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

#25306

### 💡 Background and solution

N/A

### 📝 Changelog

| Language   | Changelog                   |
| ---------- | --------------------------- |
| 🇺🇸 English | feat: support ref for Table |
| 🇨🇳 Chinese | 特性：Table 支持 ref        |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/table/index.zh-CN.md [hunk 1]
```diff
@@ -92,6 +92,8 @@ const columns = [
 | onChange | 分页、排序、筛选变化时触发 | function(pagination, filters, sorter, extra: { currentDataSource: \[], action: `paginate` \| `sort` \| `filter` }) | - |  |
 | onHeaderRow | 设置头部行属性 | function(columns, index) | - |  |
 | onRow | 设置行属性 | function(record, index) | - |  |
+| prefix | 自定义 Table 内部前缀元素 | ReactNode | - |  |
+| suffix | 自定义 Table 内部后缀元素 | ReactNode | - |  |
```
Comment (afc163) [inline components/table/index.zh-CN.md:96]: I don't think `prefix` and `suffix` are necessary, it is same:

```
{prefix}
<Table />
{suffix}
```

Comment (JounQin) [reply components/table/index.zh-CN.md:96]: That needs a wrapper element, which I'd like to avoid.

Comment (JounQin) [reply components/table/index.zh-CN.md:96]: It is used for

```tsx
<Table prefix={stikcy && <div ref={ref} className="sticky-sentinel"></div>} />
```

Comment (JounQin) [reply components/table/index.zh-CN.md:96]: @afc163 Any further discussion? The added props are just simple, but if you don't want it strongly, I can remove it in this PR.

Comment (afc163) [reply components/table/index.zh-CN.md:96]: I prefer not adding `prefix` and `suffix`.

Comment (JounQin) [reply components/table/index.zh-CN.md]: @afc163 Done!

File: components/table/Table.tsx [hunk 2]
```diff
@@ -483,7 +486,7 @@ function Table<RecordType extends object = any>(props: TableProps<RecordType>) {
     className,
   );
   return (
-    <div className={wrapperClassNames} style={style}>
+    <div ref={ref} className={wrapperClassNames} style={style}>
```
Comment (afc163) [inline components/table/Table.tsx:489]: 用户希望的 ref 可能是传给 rc-table 的？

Comment (JounQin) [reply components/table/Table.tsx:489]: 可能，但也可能不是？需要的话感觉应该使用另一个 prop，比如 `tableRef`，或者使用最外层的 ref 去查找？

Comment (JounQin) [reply components/table/Table.tsx:489]: @afc163 Please review https://github.com/react-component/table/pull/666

Comment (JounQin) [reply components/table/Table.tsx:489]: @afc163 这里你觉得应该是 `ref + tableRef` 还是 `wrapperRef + ref` ？或者其他的？

Comment (afc163) [reply components/table/Table.tsx:489]: 我不确定，可以在 #25306 问问用户需求。

Comment (afc163) [reply components/table/Table.tsx:489]: 或者和 3.x 保持一致。

Comment (JounQin) [reply components/table/Table.tsx:489]: > 或者和 3.x 保持一致。

我不太清楚 3.x 是啥样的哈

Comment (JounQin) [reply components/table/Table.tsx:489]: @afc163 

简单看了下 https://github.com/ant-design/ant-design/blob/3.26.20/components/table/Table.tsx 的代码，`StoreTable` 是 class component，用户使用 ref 取到的是 `StoreTable` instance，并不能直接访问到内部 Table 的 ref，即 [`rcTable`](https://github.com/ant-design/ant-design/blob/3.26.20/components/table/Table.tsx#L294)。

不过按照 `rcTable` 的逻辑我个人比较倾向于 `ref + rcTableRef`。

Comment (JounQin) [reply components/table/Table.tsx:489]: @afc163 Any news?

Comment (zombieJ) [reply components/table/Table.tsx:489]: 组件没 `ref` 主要还是因为纠结应该给谁……目前看给顶层 div 是 ok 的，因为 Table 没有 `focus` 和 `blur` 的需求。

**Review by afc163 — APPROVED** [review]

**Comment by Rebornnn** [comment]
<Table /> can not set Generics type

**Comment by JounQin** [comment]
> can not set Generics type

Sorry, then a new PR is needed to support it.

**Pull request merged**